torchvision/prototype/transforms/_auto_augment.py

```diff
@@ -395,4 +418,115 @@ def forward(self, *inputs: Any) -> Any:
+        if isinstance(input, (features.BoundingBox, features.SegmentationMask)):
+            raise TypeError(f"{type(input).__name__}'s are not supported by {type(self).__name__}()")
+        elif isinstance(input, torch.Tensor):
+            image = input
+        elif isinstance(input, PIL.Image.Image):
+            image = pil_to_tensor(input)
+        else:
+            return input
+
+        augmentation_space = self._AUGMENTATION_SPACE if self.all_ops else self._PARTIAL_AUGMENTATION_SPACE
+
+        image_size = F.get_image_size(image)
+
+        orig_dims = list(image.shape)
+        batch = image.view([1] * max(4 - image.ndim, 0) + orig_dims)
+        batch_dims = [batch.size(0)] + [1] * (batch.ndim - 1)
+
+        # Sample the beta weights for combining the original and augmented image. To get Beta, we use a Dirichlet
+        # with 2 parameters. The 1st column stores the weights of the original and the 2nd the ones of augmented image.
+        m = self._sample_dirichlet(
+            torch.tensor([self.alpha, self.alpha], device=batch.device).expand(batch_dims[0], -1)
+        )
+
+        # Sample the mixing weights and combine them with the ones sampled from Beta for the augmented images.
+        combined_weights = self._sample_dirichlet(
+            torch.tensor([self.alpha] * self.mixture_width, device=batch.device).expand(batch_dims[0], -1)
+        ) * m[:, 1].view([batch_dims[0], -1])
+
+        mix = m[:, 0].view(batch_dims) * batch
+        for i in range(self.mixture_width):
+            aug = batch
+            depth = self.chain_depth if self.chain_depth > 0 else int(torch.randint(low=1, high=4, size=(1,)).item())
+            for _ in range(depth):
+                transform_id, (magnitudes_fn, signed) = self._get_random_item(augmentation_space)
+
+                magnitudes = magnitudes_fn(self._PARAMETER_MAX, image_size)
+                if magnitudes is not None:
+                    magnitude = float(magnitudes[int(torch.randint(self.severity, ()))])
+                    if signed and torch.rand(()) <= 0.5:
+                        magnitude *= -1
+                else:
+                    magnitude = 0.0
+
+                aug = self._apply_transform_to_item(
+                    image, transform_id, magnitude, interpolation=self.interpolation, fill=self._parse_fill(image)
+                )
+            mix.add_(combined_weights[:, i].view(batch_dims) * aug)
+        mix = mix.view(orig_dims).to(dtype=image.dtype)
+
+        if isinstance(input, features.Image):
+            return features.Image.new_like(input, mix)
+        elif isinstance(input, torch.Tensor):
+            return mix
+        else:  # isinstance(input, PIL.Image.Image):
+            return to_pil_image(mix)
+
+    def forward(self, *inputs: Any) -> Any:
+        sample = inputs if len(inputs) > 1 else inputs[0]
+        self._query_image(sample)
+        return apply_recursively(self._apply_augmix, sample)
```

Comment:
Because the other transformations look somewhat like

```py
def forward(sample):
    for transform in transforms:
        sample = apply_recursively(transform, sample)
        
    return sample
```

whereas AugMix looks like

```py
def forward(sample):
    for item in sample:
        items = []
        for transform in transforms:
            items.append(transform(item))

        item = mix(items)

    return sample
```

IIUC, we cannot apply the transformations sequentially to the whole sample, but rather need to apply all of them to each item in the sample individually and mix the results together.

If we go the same route, we need to somehow collate multiple dictionaries. This is possible, but more complex than the current solution. The torchdata team is currently working on `default_collate` to allow hooking adding custom behavior. If that is landed, I can revisit to see if we can have a more aligned version.